src/ledger/LedgerTxnClaimableBalanceSQL.cpp

```diff
@@ -0,0 +1,399 @@
+        st.exchange(soci::use(mBalanceIDs));
+        st.exchange(soci::use(mClaimableBalanceEntrys));
+        st.exchange(soci::use(mLastModifieds));
+        st.define_and_bind();
+        {
+            auto timer = mDb.getUpsertTimer("claimablebalance");
+            st.execute(true);
+        }
+        if (static_cast<size_t>(st.get_affected_rows()) != mBalanceIDs.size())
+        {
+            throw std::runtime_error("Could not update data in SQL");
+        }
+    }
+
+    void
+    doSqliteSpecificOperation(soci::sqlite3_session_backend* sq) override
+    {
+        doSociGenericOperation();
+    }
+
+#ifdef USE_POSTGRES
+    void
+    doPostgresSpecificOperation(soci::postgresql_session_backend* pg) override
+    {
+        std::string strBalanceIDs, strClaimableBalanceEntry, strLastModifieds;
+
+        PGconn* conn = pg->conn_;
+        marshalToPGArray(conn, strBalanceIDs, mBalanceIDs);
+        marshalToPGArray(conn, strClaimableBalanceEntry,
+                         mClaimableBalanceEntrys);
+        marshalToPGArray(conn, strLastModifieds, mLastModifieds);
+
+        std::string sql =
+            "WITH r AS "
+            "(SELECT unnest(:ids::TEXT[]), unnest(:v1::TEXT[]), "
+            "unnest(:v2::INT[]))"
+            "INSERT INTO claimablebalance "
+            "(balanceid, claimablebalanceentry, lastmodified) "
+            "SELECT * FROM r "
+            "ON CONFLICT (balanceid) DO UPDATE SET "
+            "balanceid = excluded.balanceid, claimablebalanceentry = "
+            "excluded.claimablebalanceentry, "
+            "lastmodified = excluded.lastmodified";
+
+        auto prep = mDb.getPreparedStatement(sql);
+        soci::statement& st = prep.statement();
+        st.exchange(soci::use(strBalanceIDs));
+        st.exchange(soci::use(strClaimableBalanceEntry));
+        st.exchange(soci::use(strLastModifieds));
+        st.define_and_bind();
+        {
+            auto timer = mDb.getUpsertTimer("claimablebalance");
+            st.execute(true);
+        }
+        if (static_cast<size_t>(st.get_affected_rows()) != mBalanceIDs.size())
+        {
+            throw std::runtime_error("Could not update data in SQL");
+        }
+    }
+#endif
+};
+
+void
+LedgerTxnRoot::Impl::bulkUpsertClaimableBalance(
+    std::vector<EntryIterator> const& entries)
+{
+    BulkUpsertClaimableBalanceOperation op(mDatabase, entries);
+    mDatabase.doDatabaseTypeSpecificOperation(op);
+}
+
+void
+LedgerTxnRoot::Impl::dropClaimableBalances()
+{
+    throwIfChild();
+    mEntryCache.clear();
+    mBestOffersCache.clear();
+
+    std::string coll = mDatabase.getSimpleCollationClause();
+
+    mDatabase.getSession() << "DROP TABLE IF EXISTS claimablebalance;";
+    mDatabase.getSession() << "CREATE TABLE claimablebalance ("
+                           << "balanceid             VARCHAR(56) " << coll
+                           << " PRIMARY KEY, "
+                           << "claimablebalanceentry TEXT " << coll
```

Comment:
we should store not just `ClaimableBalanceEntry` but the full `LedgerEntry` in this column, this duplicates a small amount of data but avoids adding an extra column for `ledgerext` (this was added in the latest round of changes to the database)

Reply:
This would also allow us to remove the `lastmodified` column right?

Reply:
Nevermind. The `lastmodified` column is required.

Reply:
I made the change to store the full `LedgerEntry`.